expr/func_keeplastvalue_test.go

```diff
@@ -0,0 +1,199 @@
+package expr
+
+import (
+	"math"
+	"strconv"
+	"testing"
+
+	"github.com/grafana/metrictank/api/models"
+	"github.com/grafana/metrictank/test"
+	"gopkg.in/raintank/schema.v1"
+)
+
+func TestKeepLastValueAll(t *testing.T) {
+	out := []schema.Point{
+		{Val: 0, Ts: 10},
+		{Val: 0, Ts: 20},
+		{Val: 5.5, Ts: 30},
+		{Val: 5.5, Ts: 40},
+		{Val: 5.5, Ts: 50},
+		{Val: 1234567890, Ts: 60},
+	}
+
+	testKeepLastValue(
+		"keepAll",
+		math.MaxInt64,
+		[]models.Series{
+			{
+				Interval:   10,
+				Target:     "a",
+				Datapoints: getCopy(a),
+			},
+		},
+		[]models.Series{
+			{
+				Interval:   10,
+				Target:     "keepLastValue(a)",
+				Datapoints: out,
+			},
+		},
+		t,
+	)
+}
+
+func TestKeepLastValueNone(t *testing.T) {
+
+	testKeepLastValue(
+		"keepNone",
+		0,
+		[]models.Series{
+			{
+				Interval:   10,
+				Target:     "sum4a2b",
+				Datapoints: getCopy(sum4a2b),
+			},
+		},
+		[]models.Series{
+			{
+				Interval:   10,
+				Target:     "keepLastValue(sum4a2b)",
+				Datapoints: getCopy(sum4a2b),
+			},
+		},
+		t,
+	)
+}
+
+func TestKeepLastValueOne(t *testing.T) {
+	out := []schema.Point{
+		{Val: 0, Ts: 10},
+		{Val: math.MaxFloat64, Ts: 20},
+		{Val: math.MaxFloat64 - 20, Ts: 30},
+		{Val: math.MaxFloat64 - 20, Ts: 40},
+		{Val: 1234567890, Ts: 50},
+		{Val: 1234567890, Ts: 60},
+	}
+
+	testKeepLastValue(
```

Comment:
similar to feedback in previous PR, using utility functions is good, but if they call t.Fatal internally, it's harder to trace back to the calling function so if a test fails it's hard to tell where it failed.
better for the util function to return an error, and here, check for error and if so call t.Fatal.

not a huge deal though, if you have more important stuff to do, that's fine. but something to keep in mind for the future at least.

Reply:
Yeah I didn't have the utility functions when I wrote this (I just rebased), so didn't do that. Will use them in the future though.